src/main/groovy/com/morpheusdata/nutanix/prismelement/plugin/sync/ImagesSync.groovy

```diff
@@ -0,0 +1,247 @@
+ */
+@Slf4j
+class ImagesSync {
+	private final MorpheusContext context
+	private final HttpApiClient client
+	private final Cloud cloud
+	// TODO: The API restricts it to either qcow2 or iso, and then we
+	//       strip out the iso in the start of execute, so do we need an allow list?
+	private final List<String> ALLOWED_IMAGE_TYPES = ['qcow2', 'disk', 'raw', 'iso']
+
+	ImagesSync(MorpheusContext context, Cloud cloud, HttpApiClient client) {
+		this.context = context
+		this.cloud = cloud
+		this.client = client
+	}
+
+	def execute() {
+		log.info("Executing image sync for cloud $cloud.name")
+
+		try {
+			def authConfig = NutanixPrismElementPlugin.getAuthConfig(context, cloud)
+			def listResults = NutanixPrismElementApiService.listImages(client, authConfig)
+			if (listResults.success == true) {
+				//ignore isos for now
+				def cloudImages = listResults?.images?.findAll { it.imageType != 'iso' }
+
+				def existingLocations = context.async.virtualImage.location.list(
+					new DataQuery()
+						.withFilter(
+							new DataAndFilter(
+								new DataFilter('refType', 'ComputeZone'),
+								new DataFilter('refId', cloud.id),
+							)
+						)
+				)
+
+				// Sync all the the locations of the images and then within that context we'll sync the images themselves.
+				SyncTask<VirtualImageLocationIdentityProjection, Map, VirtualImageLocation> syncTask = new SyncTask<>(existingLocations, cloudImages)
+				syncTask.addMatchFunction { VirtualImageLocationIdentityProjection imageLocationProjection, Map nutanixImage ->
+					return nutanixImage.uuid == imageLocationProjection.externalId
+						|| nutanixImage?.name == imageLocationProjection.imageName
+						|| nutanixImage?.vmDiskId == imageLocationProjection.externalId
+				}.withLoadObjectDetailsFromFinder { List<SyncTask.UpdateItemDto<VirtualImageLocationIdentityProjection, VirtualImageLocation>> updateItems ->
+					context.async.virtualImage.location.listById(updateItems.collect { it.existingItem.id } as List<Long>)
+				}.onAdd {
+					log.debug("Adding missing image locations: $it")
+					addMissingVirtualImageLocations(it)
+				}.onUpdate {
+					log.debug("Updating image locations: $it")
+					updateVirtualImageLocations(it)
+				}.onDelete {
+					log.debug("Deleting image locations: $it")
+					context.services.virtualImage.location.bulkRemove(it)
+				}.start()
+
+				// If there wasn't a corresponding location for an image, remove that image
+				removeSyncedImagesWithoutLocations()
+			}
+		} catch (e) {
+			log.error("Image sync error for cloud $cloud.name: ${e}", e)
+		}
+	}
+
+	private addMissingVirtualImageLocations(Collection<Map> addItems) {
+		def existingRecords = context.async.virtualImage.listIdentityProjections(
+			new DataQuery()
+				.withFilters(
+					new DataFilter('imageType', 'in', ALLOWED_IMAGE_TYPES),
+					new DataOrFilter(
+						new DataFilter('owner', null),
+						new DataFilter('owner.id', '==', cloud.owner.id)
+					)
+				)
```

Comment:
I would probably add another dataOrFilter here for your match conditions in the match function to load the least amount of images here. The default batch size is 50, so you could be loading a lot more images than necessary.